crates/zksync/compiler/src/zksolc/compile.rs

```diff
@@ -401,40 +540,95 @@ impl ZkSolc {
+            let mut contract_artifact_paths = vec![];
+            for (idx, (contract_path, _)) in contracts.iter().enumerate() {
+                let mut artifact_path =
+                    self.project.paths.artifacts.join(PathBuf::from(&artifact_key));
+                if contracts.len() > 1 {
+                    // naming conflict where the `artifact_path` belongs to two conflicting
+                    // contracts need to adjust the paths properly
+
+                    // we keep the top most conflicting file unchanged
+                    let is_top_most = contracts.iter().enumerate().filter(|(i, _)| *i != idx).all(
+                        |(_, (cp, _))| {
+                            Path::new(contract_path).components().count() <
+                                Path::new(cp).components().count()
+                        },
+                    );
+                    if !is_top_most {
+                        // we resolve the conflicting by finding a new unique, alternative path
+                        artifact_path = Self::conflict_free_output_file(
+                            &final_artifact_paths,
+                            artifact_path,
+                            Path::new(&contract_path),
+                            &self.project.paths.artifacts,
+                        );
                     }
-                })
-                .and_then(|_| {
-                    File::open(&artifact_paths.artifact).and_then(|mut file| {
-                        let mut buffer = Vec::new();
-                        file.read_to_end(&mut buffer).map(|_| Some(buffer))
+                }
+                final_artifact_paths.insert(artifact_path.clone());
+                contract_artifact_paths.push(artifact_path);
+            }
+            for (idx, artifact_path) in contract_artifact_paths.into_iter().enumerate() {
+                contracts[idx].1.file = artifact_path
+            }
+        }
+
+        // Recreate the artifacts with their canonical file path, as generated by foundry's solc
+        // compilation
+        all_artifacts
+            .into_iter()
+            .map(|(contract_path, artifacts)| {
+                let artifacts = artifacts
+                    .into_values()
+                    .map(|artifact| {
+                        let artifact_path =
+                            artifact.file.to_str().expect("must be valid path").to_string();
+                        (artifact_path, vec![artifact])
                     })
-                })
-                .ok()
-                .flatten()
-        } else {
-            None
+                    .collect();
+                (contract_path, artifacts)
+            })
+            .collect()
+    }
+
+    /// Checks if the contract has already been compiled for the given input contract hash.
+    /// If yes, returns the pre-compiled data.
+    fn check_cache(&self, cache_path: &Path, contract_path: &Path) -> Result<CachedContractEntry> {
```

Comment:
so the `contract_path` is supposed to the the source filename _only_, right?

Reply:
`contract_path` would be the absolute filepath